spec/controllers/task_contributions_controller_spec.rb

```diff
@@ -0,0 +1,288 @@
+
+require 'rails_helper'
+
+RSpec.describe TaskContributionsController do
+  render_views
+
+  let(:user) { create(:user) }
+  let(:original_author) { create(:user) }
+  let!(:task) { create(:task, user: original_author, access_level: 'public') }
+  let(:contrib_task) { build(:task, user:, title: 'Modified title', parent_uuid: task.uuid) }
+  let(:contribution) { build(:task_contribution, suggestion: contrib_task, base: task, status: :pending) }
+
+  before { sign_in user }
+
+  describe 'GET #index' do
+    subject(:get_request) { get :index, params: {task_id: task.id} }
+
+    context 'without any task contributions' do
+      before { get_request }
+
+      expect_assigns(task: :task, task_contributions: [])
+      expect_http_status(:success)
+      expect_template(:index)
+    end
+
+    context 'with task contributions' do
+      before do
+        contribution.save!
+        get_request
+      end
+
+      expect_assigns(task: :task, task_contributions: TaskContribution.all)
+      expect_http_status(:success)
+      expect_template(:index)
+    end
+  end
+
+  describe 'GET #new' do
+    it 'assigns a new task as @task' do
+      get :new, params: {task_id: task.id}
+      expect(assigns(:task)).to be_a_new(Task)
+    end
+  end
+
+  describe 'POST #create' do
+    subject(:post_request) { post :create, params: {task_id: task.id, task: task_params} }
+
+    let(:task_params) { attributes_for(:task) }
+    let(:suggestion) { assigns(:task_contribution).suggestion }
+
+    context 'with valid params' do
+      it 'creates a new Task and TaskContribution' do
+        expect { post_request }
+          .to change(Task, :count).by(1)
+          .and change(TaskContribution, :count).by(1)
+      end
+
+      it 'assigns a newly created suggestion as @task_contribution.suggestion' do
+        post_request
+        expect(suggestion).to be_a(Task)
+        expect(suggestion).to be_persisted
+      end
+
+      it 'changes required parameters' do
+        post_request
+        expect(suggestion.access_level).to eq('private')
+        expect(suggestion.user).to eq(user)
+        expect(suggestion.parent_uuid).to eq(task.uuid)
+        expect(suggestion.task_contribution).not_to be_nil
+        expect(suggestion.task_contribution.status).to eq('pending')
+      end
+
+      it 'redirects to the created task' do
+        post_request
+        expect(response).to have_http_status(:redirect)
+        expect(response).to redirect_to([assigns(:task), assigns(:task_contribution)])
+      end
+    end
+
+    context 'when new task is invalid' do
+      let(:task_params) { {title: ''} }
+
+      it 'shows validation errors' do
+        post_request
+        expect(assigns(:task).errors.messages).to include(:title)
+      end
+
+      it 'renders the form again' do
+        post_request
+        expect(response).to render_template('tasks/new')
+      end
+    end
+  end
+
+  describe 'POST #approve_changes' do
+    let(:user) { original_author }
```

Comment:
Changing the `user` does not only have an effect on the session (allowing the `POST` to `approve_changes` to pass), but also changes the user who owns the contribution.

Specifically, in my test:
- the `task` is owned by `user`
- the `contribution` is owned by the `user`, too.

The current setup is fine to pass the test, but it is semantically difficult.

Reply:
I have added an additional variable to avoid this effect.